### PR TITLE
Password protect

### PR DESCRIPTION
**Pull request by TahaKhanAbdalli**
Password of users protected in localforage db and also password removed from being stored in localstorage as token

### CODE REVIEW

**Comment by IanMayo** [comment]
Ok, I _almost_ understand what is happening.

But, `getToken()` should return the full user details object, so we can access the `adminRights` field. We can then delete the `adminRights` property when we push the object to the cookie.

**Comment by TahaKhanAbdalli** [comment]
@IanMayo for the getToken comment. If we delete the adminRights field before pushing to the object to the cookie, then it will disturb our routing mechanism because the auth provider calls getToken and from the retun values it checks if the user has admin rights or not. if we are gonna delete that then we will not be able to determine if the user is admin or not.

**Comment by TahaKhanAbdalli** [comment]
Yes we can @IanMayo. I'll fix the warning coming in the console.

**Comment by IanMayo** [comment]
> @IanMayo for the getToken comment. If we delete the adminRights field before pushing the object to the cookie, then it will disturb our routing mechanism because the auth provider calls getToken and from the retun values it checks if the user has admin rights or not. if we are gonna delete that then we will not be able to determine if the user is admin or not.

But, if we store the whole of the `User` object in the encrypted object, we'll be able to determine the `adminUser` status, won't we?  So `getToken` would just decrypt the whole `User` object and return it.

**Comment by TahaKhanAbdalli** [comment]
Yes got it we can encrypt whole of the object. Doing it will let you know once done.

**Comment by TahaKhanAbdalli** [comment]
I am currently utilizing the bcryptjs library to encrypt and compare passwords for added security measures. This library makes use of the crypto module which is primarily designed for server-side operations, which is why the warning is being triggered. I have not come across any other library that performs this specific function. One possible solution is to apply the same encryption process as we use for local storage and subsequently decrypt and compare the user's password against the stored password in the database.

**Comment by IanMayo** [comment]
> I am currently utilizing the bcryptjs library to encrypt and compare passwords for added security measures. This library makes use of the crypto module which is primarily designed for server-side operations, which is why the warning is being triggered. I have not come across any other library that performs this specific function. One possible solution is to apply the same encryption process as we use for local storage and subsequently decrypt and compare the user's password against the stored password in the database.

Thanks Taha.  Would you mind doing a little research on this?  We don't need `bomb-proof` security, just enough to dissuade a bored staff member playing with the app.  Could you see if there is a client-side lib (or even pair of methods) that we could use with the `salt` value to encrypt/decrypt the user login details?

File: src/providers/authProvider/index.ts [hunk 1]
```diff
@@ -4,24 +4,22 @@ import { AuditType, trackEvent } from '../../utils/audit'
...
+  if (encryptedUser !== null && salt !== null) {
+    const decryptedData = decryptData(`${encryptedUser}`)
+    return decryptedData.substring(0, decryptedData.length - 32)
   } else return null
 }
 
-const setToken = (token: string): void => {
+const setToken = (token: string, salt: string): void => {
   localStorage.setItem(constants.TOKEN_KEY, token)
+  localStorage.setItem(constants.SALT, salt)
```
Comment (IanMayo) [inline src/providers/authProvider/index.ts:16]: But @TahaKhanAbdalli - doesn't thevalue of `salt` come from the environment variable?  Why are we storing it in the cookie?

File: src/providers/authProvider/index.ts [hunk 2]
```diff
@@ -2,25 +2,23 @@ import { type AuthProvider, type DataProvider } from 'react-admin'
...
-  if (JsonUser !== null) {
-    const user = JSON.parse(JsonUser) as User & { salt: string }
-    const decryptedData = decryptData(`${user.id}${user.salt}`)
-    user.id = decryptedData.substring(
-      0,
-      decryptedData.length - user.salt.length
-    )
-    return JSON.stringify(user)
-  } else return null
+export const getUser = (): string | undefined => {
```
Comment (IanMayo) [inline src/providers/authProvider/index.ts:5]: could this return an instance of `User` please?  Then the rest of the app doesn't need to know how we store the user data.

File: src/providers/authProvider/index.ts (hunk 1 above)
Comment (TahaKhanAbdalli) [reply src/providers/authProvider/index.ts:16]: It is general practice to store salt along with the encrypted data. What we are getting from env is the key.

Comment (IanMayo) [reply src/providers/authProvider/index.ts:16]: Sure, ok @TahaKhanAbdalli

File: src/providers/authProvider/index.ts (hunk 2 above)
Comment (TahaKhanAbdalli) [reply src/providers/authProvider/index.ts:5]: It was already there (returning string) but I can return user.

Comment (IanMayo) [reply src/providers/authProvider/index.ts:5]: `getUser = (): User | undefined`

**Comment by IanMayo** [comment]
For the `Codacy` errors @TahaKhanAbdalli  - you need to explicitly declare the members of the template as strings.  (or, you may be able to provide another solution)
image

**Comment by TahaKhanAbdalli** [comment]
ok I understand @IanMayo

**Comment by IanMayo** [comment]
Nice work, thanks @TahaKhanAbdalli 

There just seems one bit remaining: to be able to view/edit the password in `UserEdit`

**Comment by TahaKhanAbdalli** [comment]
> Nice work, thanks @TahaKhanAbdalli
> 
> There just seems one bit remaining: to be able to view/edit the password in `UserEdit`

Got it @IanMayo . Will work on it.

**Comment by IanMayo** [comment]
@TahaKhanAbdalli  - when I add a user, the password doesn't get encrypted:
image

I don't think this is a symptom of us temporarily having an in-memory database - the password for new users still still need to be encrypted even when we move to formal database.  Aah, but maybe that will happen on the server.

Anyway - let's get this issue sorted, then we can move on.

**Comment by TahaKhanAbdalli** [comment]
I have also encrypted password on user create.

**Pull request merged**